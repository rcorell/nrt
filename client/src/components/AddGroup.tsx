import { useMutation } from '@apollo/client';
import React, { useState } from 'react';
import { Button, Form, FormControlProps } from 'react-bootstrap';

import { createGroupMutationString } from 'src/api/api';
import { CreateGroupMutation, CreateGroupMutationVariables } from 'src/api/__generated__/CreateGroupMutation';
import { setBrowserTitle } from 'src/utils';
import { AppError, AppForm, FormContainer } from 'src/styles/form';

export const AddGroup: React.FC = () => {
    const [description, setDescription] = useState('');
    const [name, setName] = useState('');

    setBrowserTitle('Add Group');

    const [createGroup, { error }] = useMutation<CreateGroupMutation, CreateGroupMutationVariables>(
        createGroupMutationString,
        {
            variables: { description, name },
            onCompleted: ({ createGroup }) => {
                createGroup.id += 0;
            }
        }
    );

    const isFormInvalid = () => {
        return name.length < 1;
    };

    const handleSubmit = (event: React.FormEvent<FormControlProps>) => {
        event.preventDefault();

        createGroup();
    };

    const errorDisplay = () => {
        if (error) {
            return <div>Error: {JSON.stringify(error)}</div>;
        }

        return null;
    };

    return (
        <FormContainer>
            <h1>Add Group</h1>
            {errorDisplay()}
            <AppForm onSubmit={handleSubmit}>
                <Form.Group>
                    <Form.Label>Name</Form.Label>
                    <Form.Control
                        autoFocus
                        id="name"
                        onChange={(e) => setName(e.target.value)}
                        size="lg"
                        value={name}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label>Description</Form.Label>
                    <Form.Control
                        id="description"
                        onChange={(e) => setDescription(e.target.value)}
                        size="lg"
                        value={description}
                    />
                </Form.Group>

                <Button type="submit" disabled={isFormInvalid()}>
                    Add Group
                </Button>
            </AppForm>
            <AppError></AppError>
        </FormContainer>
    );
};
