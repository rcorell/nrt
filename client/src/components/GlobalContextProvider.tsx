import React, { useState } from 'react';

import { isAuthenticated } from 'src/api/api';

type GlobalContextType = {
    authenticated: boolean;
    setAuthenticated: (authenticated: boolean) => void;
};

export const GlobalContext = React.createContext<GlobalContextType>({
    authenticated: false,
    setAuthenticated: (_authenticated) => {}
});

export const GlobalContextProvider: React.FC = ({ children }) => {
    const [authenticated, setAuthenticated] = useState(isAuthenticated());

    const contextFunctions = { authenticated, setAuthenticated };

    return <GlobalContext.Provider value={contextFunctions}>{children}</GlobalContext.Provider>;
};
