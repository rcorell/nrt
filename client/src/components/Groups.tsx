import React from 'react';
import * as _ from 'lodash';
import { useMutation, useQuery } from '@apollo/client';
import MaterialTable from 'material-table';

import { fetchGroupsQueryString, fetchUserQueryString, joinGroupMutationString } from 'src/api/api';
import { FetchGroupsQuery } from 'src/api/__generated__/FetchGroupsQuery';
import { FetchUserQuery } from 'src/api/__generated__/FetchUserQuery';
import { JoinGroupMutation, JoinGroupMutationVariables } from 'src/api/__generated__/JoinGroupMutation';
import { setBrowserTitle } from 'src/utils';

// const renderJoinColumn = (rowData: any) => {
//     if (userGroups.contains(rowData.name)) {
//         return <p>Joined</p>;
//     }

//     return <p>Join</p>;
// };

const COLUMN_DEFINITIONS = [
    { title: 'Name', field: 'name' },
    { title: 'Description', field: 'description' },
    {
        field: 'join',
        title: ''
    }
];

export const Groups: React.FC = () => {
    setBrowserTitle('Groups');

    const userQuery = useQuery<FetchUserQuery>(fetchUserQueryString);
    const allGroups = useQuery<FetchGroupsQuery>(fetchGroupsQueryString);

    const [joinGroup] = useMutation<JoinGroupMutation, JoinGroupMutationVariables>(joinGroupMutationString, {
        onCompleted: () => {
            // console.log('woo');
            userQuery.refetch();
        },
        onError: () => {
            // https://github.com/apollographql/apollo-client/issues/7167
        }
    });

    if (allGroups.loading || userQuery.loading) {
        return <div>Loading...</div>;
    }

    if (allGroups.error || userQuery.error) {
        return <div>Error: {JSON.stringify(allGroups.error) + JSON.stringify(userQuery.error)}</div>;
    }

    const userGroupIds = userQuery!.data!.user.groups.map((group) => {
        return group!.id;
    });

    const groups = _.cloneDeep(allGroups.data!.groups).map((group) => {
        return {
            ...group,
            join: userGroupIds.includes(group.id) ? (
                'Joined'
            ) : (
                <button onClick={() => joinGroup({ variables: { groupId: Number(group.id) } })}>Join!</button>
            )
        };
    });

    return <MaterialTable title="Groups" columns={COLUMN_DEFINITIONS} data={groups} />;
};
