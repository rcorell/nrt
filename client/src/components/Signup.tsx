import { useMutation } from '@apollo/client';
import { navigate } from 'hookrouter';
import React, { useState } from 'react';
import { Button, Form, FormControlProps } from 'react-bootstrap';

import { signupMutationString } from 'src/api/api';
import { SignupMutation, SignupMutationVariables } from 'src/api/__generated__/SignupMutation';
import { setBrowserTitle } from 'src/utils';
import { AppError, AppForm, FormContainer } from 'src/styles/form';

export const Signup: React.FC = () => {
    const [email, setEmail] = useState('');
    const [name, setName] = useState('');
    const [password, setPassword] = useState('');

    const [signup, { error }] = useMutation<SignupMutation, SignupMutationVariables>(signupMutationString, {
        variables: { email, name, password },
        onCompleted: ({ signup }) => {
            if (signup?.token) {
                console.log(signup);
                localStorage.setItem('token', signup.token);
                navigate('/');
            }
        }
    });

    setBrowserTitle('Sign Up');

    const isFormInvalid = () => {
        return email.length < 7 || name.length === 0 || password.length === 0;
    };

    const signupStatus = () => {
        if (error) {
            return <AppError>Error: {error.message}</AppError>;
        }

        return null;
    };

    const handleSubmit = (event: React.FormEvent<FormControlProps>) => {
        event.preventDefault();

        signup();
    };

    return (
        <FormContainer>
            <h1>Sign Up</h1>
            <AppForm onSubmit={handleSubmit}>
                <Form.Group>
                    <Form.Label id="emailLabel">Email</Form.Label>
                    <Form.Control
                        autoFocus
                        aria-labelledby="emailLabel"
                        id="email"
                        onChange={(e) => setEmail(e.target.value)}
                        size="lg"
                        type="email"
                        value={email}
                        placeholder={'email'}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label id="nameLabel">Name</Form.Label>
                    <Form.Control
                        autoFocus
                        aria-labelledby="nameLabel"
                        id="name"
                        onChange={(e) => setName(e.target.value)}
                        size="lg"
                        value={name}
                        placeholder={'name'}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label id="passwordLabel">Password</Form.Label>
                    <Form.Control
                        id="password"
                        aria-labelledby="passwordLabel"
                        onChange={(e) => setPassword(e.target.value)}
                        size="lg"
                        type="password"
                        value={password}
                        placeholder={'password'}
                    />
                </Form.Group>

                <Button type="submit" disabled={isFormInvalid()}>
                    Sign Up
                </Button>
            </AppForm>
            <AppError>{signupStatus()}</AppError>
        </FormContainer>
    );
};
