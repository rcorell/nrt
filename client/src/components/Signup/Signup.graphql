mutation Signup($email: String!, $name: String!, $password: String!) {
    signup(email: $email, name: $name, password: $password) {
        token
    }
}
