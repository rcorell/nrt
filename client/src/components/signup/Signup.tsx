import * as React from 'react';
import { Button, Form, FormControlProps } from 'react-bootstrap';
import { connect } from 'react-redux';

import { State } from '../../redux/state.types';
import * as signupActions from '../../redux/signup/actions';

import { FormContainer, AuthError, AuthForm } from '../../styles';

interface SignupStateProps {
}

interface SignupDispatchProps {
    signup: (email: string, name: string, password: string) => void;
}

export type SignupComponentProps = SignupStateProps & SignupDispatchProps;

interface SignupState {
    email: string;
	name: string;
    password: string;
}

export class SignupComponent extends React.Component<SignupComponentProps, SignupState> {
    public constructor(props: SignupComponentProps) {
        super(props);

        this.state = {
            email: '',
			name: '',
            password: ''
        };
    }

    componentDidMount() {
        document.title = 'Sign Up';
    }

    isFormInvalid = () => {
        return this.state.email.length < 7 || this.state.name.length === 0 || this.state.password.length === 0;
    };

    handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        const newState: SignupState = this.state;

        switch (event.target.id) {
            case 'email':
                newState.email = event.target.value;
                break;

			case 'name':
				newState.name = event.target.value;
				break;
	
            case 'password':
                newState.password = event.target.value;
                break;
        }
        this.setState(newState);
    };

    handleSubmit = (event: React.FormEvent<FormControlProps>) => {
        event.preventDefault();
        this.props.signup(this.state.email, this.state.name, this.state.password);
    };

    render() {
        return (
            <FormContainer>
                <h1>Signup</h1>
                <AuthForm onSubmit={this.handleSubmit}>
                    <Form.Group>
                        <Form.Label>Email</Form.Label>
                        <Form.Control
                            autoFocus
                            id="email"
                            onChange={this.handleChange}
                            size="lg"
                            type="email"
                            value={this.state.email}
                            placeholder={'email'}
                        />
                    </Form.Group>
                    
                    <Form.Group>
                        <Form.Label>Name</Form.Label>
                        <Form.Control
                            autoFocus
                            id="name"
                            onChange={this.handleChange}
                            size="lg"
                            type="name"
                            value={this.state.name}
                            placeholder={'name'}
                        />
                    </Form.Group>

                    <Form.Group>
                        <Form.Label>Password</Form.Label>
                        <Form.Control
                            id="password"
                            onChange={this.handleChange}
                            size="lg"
                            type="password"
                            value={this.state.password}
                            placeholder={'password'}
                        />
                    </Form.Group>

                    <Button type="submit" disabled={this.isFormInvalid()}>
                        Log In
                    </Button>
                </AuthForm>
                <AuthError>
                </AuthError>
            </FormContainer>
        );
    }
}

export const mapStateToProps: (state: State) => SignupStateProps = (state: State) => ({
});

export const mapDispatchToProps = (dispatch: Function) => ({
    signup: (email: string, name: string, password: string) => dispatch(signupActions.signup(email, name, password))
});

export const Signup = connect(mapStateToProps, mapDispatchToProps)(SignupComponent);
