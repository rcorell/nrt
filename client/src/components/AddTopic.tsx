import { useMutation, useQuery } from '@apollo/client';
import React, { useState } from 'react';
import { Button, Form, FormControlProps } from 'react-bootstrap';

import { CreateTopicMutation, CreateTopicMutationVariables } from 'src/api/__generated__/CreateTopicMutation';
import { FetchUserQuery } from 'src/api/__generated__/FetchUserQuery';
import { createTopicMutationString, fetchUserQueryString } from 'src/api/api';
import { LOADING_TEXT } from 'src/components/shared';
import { AppError, AppForm, FormContainer } from 'src/styles/form';
import { setBrowserTitle } from 'src/utils';

export const AddTopic: React.FC = () => {
    const [description, setDescription] = useState('');
    const [groupId, setGroupId] = useState('');
    const [title, setTitle] = useState('');

    const userQuery = useQuery<FetchUserQuery>(fetchUserQueryString);

    const [createTopic, { error }] = useMutation<CreateTopicMutation, CreateTopicMutationVariables>(
        createTopicMutationString,
        {
            variables: { groupId, title, description },
            onError: () => {
                // RTL bug
            }
        }
    );

    const isFormInvalid = () => {
        return title.length < 1;
    };

    const handleSubmit = (event: React.FormEvent<FormControlProps>) => {
        event.preventDefault();

        createTopic();
    };

    const errorStatus = () => {
        if (error) {
            return <div>Error: ${JSON.stringify(error)}</div>;
        }

        return null;
    };

    setBrowserTitle('Add Topic');

    if (userQuery.loading || !userQuery.data) {
        return <div>{LOADING_TEXT}</div>;
    }

    const groupOptions = userQuery.data.user.groups.map((group) => (
        <option key={group!.id} value={group!.id}>
            {group!.name}
        </option>
    ));

    return (
        <FormContainer>
            <h1>Add Topic</h1>
            {errorStatus()}
            <AppForm onSubmit={handleSubmit}>
                <Form.Group>
                    <Form.Label id="groupLabel">Group</Form.Label>
                    <Form.Control
                        as="select"
                        aria-labelledby="groupLabel"
                        id="group"
                        onChange={(e) => setGroupId(e.target.value)}
                        size="lg"
                    >
                        {groupOptions}
                    </Form.Control>
                </Form.Group>

                <Form.Group>
                    <Form.Label id="topicLabel">Topic</Form.Label>
                    <Form.Control
                        autoFocus
                        aria-labelledby="topicLabel"
                        id="topic"
                        onChange={(e) => setTitle(e.target.value)}
                        size="lg"
                        value={title}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label id="descriptionLabel">Description</Form.Label>
                    <Form.Control
                        id="description"
                        aria-labelledby="descriptionLabel"
                        onChange={(e) => setDescription(e.target.value)}
                        size="lg"
                        value={description}
                    />
                </Form.Group>

                <Button type="submit" disabled={isFormInvalid()}>
                    Submit Topic
                </Button>
            </AppForm>
            <AppError></AppError>
        </FormContainer>
    );
};
