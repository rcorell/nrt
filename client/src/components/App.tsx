import 'bootstrap/dist/css/bootstrap.min.css';

import { ApolloProvider } from '@apollo/client';
import React from 'react';

import { GlobalContextProvider } from 'src/components/GlobalContextProvider';
import { apolloClient } from 'src/api/api';
import { Routes } from 'src/components/Routes';
import { AppHeader, GlobalStyle } from 'src/styles';

export const App: React.FC = () => {
    return (
        <>
            <GlobalStyle />
            <GlobalContextProvider>
                <ApolloProvider client={apolloClient}>
                    <AppHeader>Top 5 Daily</AppHeader>
                    <Routes />
                </ApolloProvider>
            </GlobalContextProvider>
        </>
    );
};
