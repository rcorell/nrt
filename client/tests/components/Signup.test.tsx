import { fireEvent, screen } from '@testing-library/react';

import { Signup } from 'src/components/Signup';

import { INVALID, VALID } from 'tests/fixtures';
import { renderComponent } from 'tests/testHelpers';

describe('Signup', () => {
    let signupContainer: HTMLElement;

    const setFields = (email: string, name: string, password: string) => {
        const emailInput = screen.getByLabelText('Email');
        const nameInput = screen.getByLabelText('Name');
        const passwordInput = screen.getByLabelText('Password');

        fireEvent.change(emailInput, { target: { value: email } });
        fireEvent.change(nameInput, { target: { value: name } });
        fireEvent.change(passwordInput, { target: { value: password } });
    };

    const renderSignup = () => {
        signupContainer = renderComponent(Signup).container;
    };

    describe('snapshots', () => {
        beforeEach(() => {
            renderSignup();
        });

        it('fresh load', () => {
            expect(signupContainer).toMatchSnapshot();
        });

        it('invalid parameters', () => {
            setFields(INVALID.EMAIL, INVALID.NAME, INVALID.PASSWORD);

            expect(signupContainer).toMatchSnapshot();
        });

        it('valid parameters', async () => {
            setFields(VALID.EMAIL, VALID.NAME, VALID.PASSWORD);

            expect(signupContainer).toMatchSnapshot();
        });
    });
});
