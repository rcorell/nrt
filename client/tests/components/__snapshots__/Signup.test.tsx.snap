// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup snapshots fresh load 1`] = `
<div>
  <div
    class="sc-bdvvaa hMSAon"
  >
    <h1>
      Sign Up
    </h1>
    <form
      class="sc-gsDJrp idIFaR"
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="emailLabel"
        >
          Email
        </label>
        <input
          aria-labelledby="emailLabel"
          class="form-control form-control-lg"
          id="email"
          placeholder="email"
          type="email"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="nameLabel"
        >
          Name
        </label>
        <input
          aria-labelledby="nameLabel"
          class="form-control form-control-lg"
          id="name"
          placeholder="name"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="passwordLabel"
        >
          Password
        </label>
        <input
          aria-labelledby="passwordLabel"
          class="form-control form-control-lg"
          id="password"
          placeholder="password"
          type="password"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        disabled=""
        type="submit"
      >
        Sign Up
      </button>
    </form>
    <div
      class="sc-dkPtyc cWQCqc"
    />
  </div>
</div>
`;

exports[`Signup snapshots invalid parameters 1`] = `
<div>
  <div
    class="sc-bdvvaa hMSAon"
  >
    <h1>
      Sign Up
    </h1>
    <form
      class="sc-gsDJrp idIFaR"
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="emailLabel"
        >
          Email
        </label>
        <input
          aria-labelledby="emailLabel"
          class="form-control form-control-lg"
          id="email"
          placeholder="email"
          type="email"
          value="a@a.a"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="nameLabel"
        >
          Name
        </label>
        <input
          aria-labelledby="nameLabel"
          class="form-control form-control-lg"
          id="name"
          placeholder="name"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="passwordLabel"
        >
          Password
        </label>
        <input
          aria-labelledby="passwordLabel"
          class="form-control form-control-lg"
          id="password"
          placeholder="password"
          type="password"
          value=""
        />
      </div>
      <button
        class="btn btn-primary"
        disabled=""
        type="submit"
      >
        Sign Up
      </button>
    </form>
    <div
      class="sc-dkPtyc cWQCqc"
    />
  </div>
</div>
`;

exports[`Signup snapshots valid parameters 1`] = `
<div>
  <div
    class="sc-bdvvaa hMSAon"
  >
    <h1>
      Sign Up
    </h1>
    <form
      class="sc-gsDJrp idIFaR"
    >
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="emailLabel"
        >
          Email
        </label>
        <input
          aria-labelledby="emailLabel"
          class="form-control form-control-lg"
          id="email"
          placeholder="email"
          type="email"
          value="totally@valid.email"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="nameLabel"
        >
          Name
        </label>
        <input
          aria-labelledby="nameLabel"
          class="form-control form-control-lg"
          id="name"
          placeholder="name"
          value="Buckaroo Banzai"
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class="form-label"
          id="passwordLabel"
        >
          Password
        </label>
        <input
          aria-labelledby="passwordLabel"
          class="form-control form-control-lg"
          id="password"
          placeholder="password"
          type="password"
          value="qwerty1!"
        />
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Sign Up
      </button>
    </form>
    <div
      class="sc-dkPtyc cWQCqc"
    />
  </div>
</div>
`;
